pkg/queuemanager/manager.go

```
@@ -0,0 +1,218 @@
+}
+
+type kindQueue struct {
+	queue         chan *queuedRequest
+	pool          *ants.Pool
+	maxObjectSize int
+}
+
+type QueueManager struct {
+	queues           map[string]*kindQueue
+	lastQueueFullLog map[string]time.Time
+	mu               sync.Mutex
+	cfg              *config.Config
+}
+
+func NewQueueManager(cfg *config.Config) *QueueManager {
+	return &QueueManager{
+		queues:           make(map[string]*kindQueue),
+		cfg:              cfg,
+		lastQueueFullLog: make(map[string]time.Time),
+	}
+}
+
+func GoID() int {
+	var buf [64]byte
+	n := runtime.Stack(buf[:], false)
+	// Stack trace looks like: "goroutine 12345 ["
+	fields := bytes.Fields(buf[:n])
+	if len(fields) < 2 {
+		return -1
+	}
+	id, err := strconv.Atoi(string(fields[1]))
+	if err != nil {
+		return -1
+	}
+	return id
+}
+
+func (qm *QueueManager) getOrCreateQueue(kind string) *kindQueue {
+	qm.mu.Lock()
+	defer qm.mu.Unlock()
+	q, exists := qm.queues[kind]
+	if !exists {
+		// Get per-kind config or use defaults
+		kcfg, ok := qm.cfg.KindQueues[kind]
+		queueLen := qm.cfg.DefaultQueueLength
+		workerCount := qm.cfg.DefaultWorkerCount
+		maxObjectSize := qm.cfg.DefaultMaxObjectSize
+		if ok {
+			if kcfg.QueueLength > 0 {
+				queueLen = kcfg.QueueLength
+			}
+			if kcfg.WorkerCount > 0 {
+				workerCount = kcfg.WorkerCount
+			}
+			if kcfg.MaxObjectSize > 0 {
+				maxObjectSize = kcfg.MaxObjectSize
+			}
+		}
+		queue := make(chan *queuedRequest, queueLen)
+		pool, _ := ants.NewPool(workerCount)
+		q = &kindQueue{
+			queue:         queue,
+			pool:          pool,
+			maxObjectSize: maxObjectSize,
+		}
+		logger.L().Info("queue configuration", helpers.String("kind", kind), helpers.Int("queueLength", queueLen), helpers.Int("workerCount", workerCount), helpers.Int("maxObjectSize", maxObjectSize))
+		// Start a dispatcher goroutine for this kind
+		go func(q *kindQueue) {
+			for req := range q.queue {
+				reqLocal := req // capture loop variable
+				err := q.pool.Submit(func() {
```

do you know that this type of queue is blocking if the number of workers is exceeded?

---

In this implementation, the dispatcher goroutine that reads from the queue and submits work to the worker pool can indeed be blocked if the queue is full and all workers are busy. However, in the HTTP handler (QueueHandler), we use a select statement with a default case when attempting to enqueue a request:

```
select {
case q.queue <- req:
    <-req.done
default:
    // handle queue full (e.g., return error to client)
}
```

This means that if the queue is full, the handler will not block; instead, the default case is executed. 

This allows us to fail fast and return an appropriate error response to the client (such as HTTP 429 Too Many Requests), rather than blocking the HTTP request and potentially exhausting server resources.

While the dispatcher can be blocked if the queue is full, the HTTP request handling path is protected from blocking by using the non-blocking select pattern. This ensures that the server remains responsive under load and can handle situations where the queue is at full.